### PR TITLE
Patch to improve multithreaded performance scaling

### PR DESCRIPTION
**Pull request by hcho3**
Change parallel strategy for histogram construction.
Instead of partitioning data rows among multiple threads, partition feature columns instead. Useful heuristics for assigning partitions have been adopted from [LightGBM project](https://github.com/Microsoft/LightGBM/blob/master/src/io/dataset.cpp). In particular, features that do not have overlapping nonzeros are grouped together.

How to use: by default, the new parallel strategy will be used. To revert to old behavior, simply set `enable_feature_grouping=0`.

Background: poor performance scaling with respect to number of worker threads has been reported by [a user benchmark](https://github.com/Laurae2/gbt_benchmarks). OpenMP inserts an implicit synchronization barrier at end of each parallel loop; with high work imbalance, most threads were doing no useful work but waiting. Partitioning data column-wise reduces work imbalance and thus improve performance.

### CODE REVIEW

**Comment by hcho3** [comment]
This patch may address issue #2165.

File: src/tree/param.h [hunk 1]
```diff
@@ -99,26 +90,6 @@ struct TrainParam : public dmlc::Parameter<TrainParam> {
         .describe(
             "Maximum depth of the tree; 0 indicates no limit; a limit is required "
             "for depthwise policy");
-    DMLC_DECLARE_FIELD(max_leaves).set_lower_bound(0).set_default(0).describe(
-        "Maximum number of leaves; 0 indicates no limit.");
-    DMLC_DECLARE_FIELD(max_bin).set_lower_bound(2).set_default(256).describe(
```
Comment (RAMitchell) [inline src/tree/param.h]: Removing the 'max_bins' parameter will break gpu algorithms.

Comment (hcho3) [inline src/tree/param.h]: Oops, my bad. I had erroneously assumed that the parameter was only used by the fast histogram updater. I've put back `max_bins` and other parameters back to `TrainParam`.

File: src/common/hist_util.cc [hunk 2]
```diff
@@ -154,40 +156,264 @@ void GHistIndexMatrix::Init(DMatrix* p_fmat) {
...
+          return max_cnt + 1;
+        }
+      }
+    }
+  }
+  return ret;
+}
+
+template <typename T>
+static inline void
```
Comment (tqchen) [inline src/common/hist_util.cc]: statoc keyword is not needed for inline code

**Comment by pseudotensor** [comment]
Just a note that I'm getting random crashes/hangs with the lightgbm mode after this commit.  Normal hist and exact and all gpu stuff seems unaffected.  I'll try to isolate to be more useful.

**Comment by hcho3** [comment]
@pseudotensor Thanks for letting us know. Can you check the memory usage? I now realize that the LightGBM mode is not as memory efficient as it can be.

**Comment by pseudotensor** [comment]
Memory usage isn't an issue -- very small data set.  Sorry for being vague, will try to isolate myself.

**Comment by hcho3** [comment]
@pseudotensor For now, I've submitted a patch to revert the default to old algorithm. If possible, can you send me a link for the dataset you are using?

**Comment by pseudotensor** [comment]
Hi @hcho3 , I looked more and I can't quite track down the issue.  I'll need more time.  @RAMitchell suggested, however, that things run about 3X slower.  Has he communicated about this?

**Comment by hcho3** [comment]
@pseudotensor Let me spend more time on my end as well. Even disabling feature grouping doesn't seem to solve all problems, as quantile extractions and gmat construction prove to be more costly than previously thought.

**Pull request merged**